### PR TITLE
Add V11 Support

### PR DESCRIPTION
**Pull request by bryantgillespie**
- Updated to work with V11 access policies
- Better error handling and write logs to files for easier sharing
- Programmatic mode for apply and extract commands
- Partial template application through flags

Please see README for other details and changes.

A beta version is now published and you can test it using @beta tag
`npx directus-template-cli@beta apply`
or
`npx directus-template-cli@beta extract`

### CODE REVIEW

**Comment by alexvdvalk** [comment]
I hit the pressure limiter when doing an extract from cloud.  I think the retry inside bottleneck isnt working because we catch the errors.

**Comment by arladmin** [comment]
Issues with `0.5.0-beta.15` :
https://github.com/directus-labs/directus-template-cli/issues/60#issuecomment-2351006978

**Comment by arladmin** [comment]
Is the `0.5.0-beta.15` version functional already?

Because, I'm still seeing the errors reported here: https://github.com/directus-labs/directus-template-cli/issues/60#issuecomment-2351006978

**Comment by bryantgillespie** [comment]
> Is the `0.5.0-beta.15` version functional already?
> 
> Because, I'm still seeing the errors reported here: [#60 (comment)](https://github.com/directus-labs/directus-template-cli/issues/60#issuecomment-2351006978)

This should be resolved in `0.5.0-beta.16`

**Comment by arladmin** [comment]
> > Is the `0.5.0-beta.15` version functional already?
> > Because, I'm still seeing the errors reported here: [#60 (comment)](https://github.com/directus-labs/directus-template-cli/issues/60#issuecomment-2351006978)
> 
> This should be resolved in `0.5.0-beta.16`

Hi @bryantgillespie 

I'm still seeing these errors:

image

**Comment by arladmin** [comment]
> > Is the `0.5.0-beta.15` version functional already?
> > Because, I'm still seeing the errors reported here: [#60 (comment)](https://github.com/directus-labs/directus-template-cli/issues/60#issuecomment-2351006978)
> 
> This should be resolved in `0.5.0-beta.16`

Hi @bryantgillespie 

Still seeing a bunch of errors during the `apply` process ->

image

**Comment by alexvdvalk** [comment]
We're working on it.

My recent commit should resolve the "Value is Required" errors

**Comment by arladmin** [comment]
Hi @bryantgillespie  @alexvdvalk 

Again, the same issue with the latest beta (`0.5.0-beta.17`).

image

```
-- Logged in as Admin User
=== Applying template - test4 to https://*************

Loading 88 collections and 819 fields... done
Loading 209 relations... done
Loading 6 roles... done
Loading 10 policies... done
Loading 435 permissions... done
Loading 20 users... done
 »   Warning: Directus Error: Invalid foreign key for field "role" in collection "directus_access". (Status: 400) {"collection":"directus_access","field":"role","code":"INVALID_FOREIGN_KEY"}
 »   Context: {"access":{"id":"ac307b5c-623e-40c0-9456-34a48d9da283","role":"dc8a1314-b911-49e7-a90d-3273e1767fb9","user":null,"policy":"dc8a1314-b911-49e7-a90d-3273e1767fb9","sort":1},"operation":"createAccess"}
Loading 10 accesses... done
Loading 29 folders... done
Loading 228 files... done
 »   Warning: Directus Error: Validation failed for field "title". Value is required. (Status: 400) {"field":"title","type":"required","code":"FAILED_VALIDATION"}
 »   Warning: Directus Error: Validation failed for field "component". Value is required. (Status: 400) {"field":"component","type":"required","code":"FAILED_VALIDATION"}
 »   Warning: Directus Error: Validation failed for field "component". Value is required. (Status: 400) {"field":"component","type":"required","code":"FAILED_VALIDATION"}
 »   Warning: Directus Error: Validation failed for field "title". Value is required. (Status: 400) {"field":"title","type":"required","code":"FAILED_VALIDATION"}
 »   Warning: Directus Error: Validation failed for field "status". Value is required. (Status: 400) {"field":"status","type":"required","code":"FAILED_VALIDATION"}
 »   Warning: Directus Error: Validation failed for field "title". Value is required. (Status: 400) {"field":"title","type":"required","code":"FAILED_VALIDATION"}
 »   Warning: Directus Error: Validation failed for field "title". Value is required. (Status: 400) {"field":"title","type":"required","code":"FAILED_VALIDATION"}
 »   Warning: Directus Error: Validation failed for field "title". Value is required. (Status: 400) {"field":"title","type":"required","code":"FAILED_VALIDATION"}
 »   Warning: Directus Error: Validation failed for field "slug". Value is required. (Status: 400) {"field":"slug","type":"required","code":"FAILED_VALIDATION"}
 »   Warning: Directus Error: Validation failed for field "email". Value is required. (Status: 400) {"field":"email","type":"required","code":"FAILED_VALIDATION"}
-- Request failed. Retrying after 1000ms
-- Retrying job (attempt 1)
 »   Warning: Directus Error: Invalid foreign key for field "platform_id" in collection "marketing_post_platform". (Status: 400) {"collection":"marketing_post_platform","field":"platform_id","code":"INVALID_FOREIGN_KEY"}
 »   Warning: Directus Error: Invalid foreign key for field "post_id" in collection "marketing_post_platform". (Status: 400) {"collection":"marketing_post_platform","field":"post_id","code":"INVALID_FOREIGN_KEY"}
 »   Warning: Directus Error: Invalid foreign key for field "post_id" in collection "marketing_post_platform". (Status: 400) {"collection":"marketing_post_platform","field":"post_id","code":"INVALID_FOREIGN_KEY"}
 »   Warning: Directus Error: Invalid foreign key for field "post_id" in collection "marketing_post_tags". (Status: 400) {"collection":"marketing_post_tags","field":"post_id","code":"INVALID_FOREIGN_KEY"}
 »   Warning: Directus Error: Invalid foreign key for field "tags_id" in collection "marketing_post_tags". (Status: 400) {"collection":"marketing_post_tags","field":"tags_id","code":"INVALID_FOREIGN_KEY"}
 »   Warning: Directus Error: Invalid foreign key for field "marketing_tags_id" in collection "marketing_tags_platform". (Status: 400) {"collection":"marketing_tags_platform","field":"marketing_tags_id","code":"INVALID_FOREIGN_KEY"}
-- Request failed. Retrying after 1000ms
-- Retrying job (attempt 1)
-- Request failed. Retrying after 1000ms
-- Retrying job (attempt 1)
 »   Warning: Directus Error: Invalid foreign key for field "posts_id" in collection "post_gallery_items". (Status: 400) {"collection":"post_gallery_items","field":"posts_id","code":"INVALID_FOREIGN_KEY"}
-- Request failed. Retrying after 1000ms
-- Retrying job (attempt 1)
-- Request failed. Retrying after 1000ms
-- Retrying job (attempt 1)
-- Request failed. Retrying after 1000ms
-- Retrying job (attempt 1)
-- Request failed. Retrying after 1000ms
-- Retrying job (attempt 1)
 »   Warning: Directus Error: Invalid foreign key for field "posts_id" in collection "post_gallery_items". (Status: 400) {"collection":"post_gallery_items","field":"posts_id","code":"INVALID_FOREIGN_KEY"}
-- Request failed. Retrying after 2000ms
-- Retrying job (attempt 2)
-- Request failed. Retrying after 2000ms
-- Retrying job (attempt 2)
-- Request failed. Retrying after 2000ms
-- Retrying job (attempt 2)
-- Request failed. Retrying after 2000ms
-- Retrying job (attempt 2)
-- Request failed. Retrying after 1000ms
-- Retrying job (attempt 1)
-- Request failed. Retrying after 4000ms
-- Retrying job (attempt 3)
-- Request failed. Retrying after 4000ms
-- Retrying job (attempt 3)
-- Request failed. Retrying after 4000ms
-- Retrying job (attempt 3)
-- Request failed. Retrying after 4000ms
-- Retrying job (attempt 3)
-- Max retries reached, not retrying further
 »   Warning: Directus Error: You don't have permission to access this. (Status: 403) {"code":"FORBIDDEN"}
-- Max retries reached, not retrying further
 »   Warning: Directus Error: You don't have permission to access this. (Status: 403) {"code":"FORBIDDEN"}
-- Max retries reached, not retrying further
 »   Warning: Directus Error: You don't have permission to access this. (Status: 403) {"code":"FORBIDDEN"}
-- Max retries reached, not retrying further
 »   Warning: Directus Error: You don't have permission to access this. (Status: 403) {"code":"FORBIDDEN"}
 »   Warning: Directus Error: Invalid foreign key for field "featured_post" in collection "pages_blog". (Status: 400) {"collection":"pages_blog","field":"featured_post","code":"INVALID_FOREIGN_KEY"}
Loading data for 88 collections... done
Updating 36 fields to required... done
Loading 4 dashboards... done
Loading 26 flows... done
Loading settings... done
Loading 0 translations... done
Loading 65 presets... done
Found 5 extensions total: 3 registry extensions (including 0 bundles), and 2 local extensions
-- Installed directus-extension-editable-layout
-- Installed directus-extension-display-link
-- Installed directus-extension-api-trigger-interface
Installing 3 extensions... done
Finished installing extensions
Note: 2 local extensions need to be installed manually.
------------------
Template applied successfully.
```

**Comment by alexvdvalk** [comment]
@arladmin  Are you able to compress your template and share it with us?

**Comment by arladmin** [comment]
> @arladmin Are you able to compress your template and share it with us?

@alexvdvalk , i can, but only privately

**Comment by arladmin** [comment]
Hi @alexvdvalk 

Do let me know where/how to share the Directus project with you?

**Comment by alexvdvalk** [comment]
Can you provide an email where I can contact you?

**Comment by arladmin** [comment]
> Can you provide an email where I can contact you?

@alexvdvalk 

Let me contact you via Discord (would like to avoid immortalizing my email here :D )

**Comment by sandias42** [comment]
Update: I tried `npx directus-template-cli@0.5.0-beta.17 apply` to repeat the agencyOS template installation and it looks like  `policies.json` hasn't been updated yet, possibly you do this extraction in CI/CD or something? I'll be looking around for a solve for the next hour. Would very much appreciate pointers for a work around which could make agencyOS usable while you are finalizing this fix :)

```
Loading 76 collections and 700 fields... done
Loading 176 relations... done
Loading 4 roles... done
 ›   Warning: Unknown error: "File not found: 
 ›   /home/runner/.npm/_npx/081441f1bf72b51a/node_modules/directus-template-cli/dist/lib/downloads/official/agencyos/src/policies.json"
    Error: ENOENT: no such file or directory, open 
    '/home/runner/.npm/_npx/081441f1bf72b51a/node_modules/directus-template-cli/dist/lib/downloads/official/agencyos/src/policies.json'
    Code: ENOENT
    ```

**Comment by arladmin** [comment]
These are the error logs for the `apply` method with the latest 'beta' (directus-template-cli@0.5.0-beta.18)  ->
```
-- Logged in as Admin User
=== Applying template - test8 to https://my-domain/

Loading 54 collections and 524 fields... done
Loading 122 relations... done
Loading 5 roles... done
Loading 6 policies... done
Loading 290 permissions... done
Loading 19 users... done
 »   Warning: Directus Error: Invalid foreign key for field "role" in collection "directus_access". (Status: 400) {"collection":"directus_access","field":"role","code":"INVALID_FOREIGN_KEY"}
 »   Context: {"access":{"id":"3107a8ed-18ea-44fa-ac19-866536aac9f0","role":"c8cc6feb-db01-40a5-bbdc-ead36a040115","user":null,"policy":"c8cc6feb-db01-40a5-bbdc-ead36a040115","sort":1},"operation":"createAccess"}
Loading 6 accesses... done
Loading 27 folders... done
Loading 235 files... done
-- Request failed. Retrying after 1000ms
-- Retrying job (attempt 1)
-- Request failed. Retrying after 1000ms
-- Retrying job (attempt 1)
-- Request failed. Retrying after 1000ms
-- Retrying job (attempt 1)
-- Request failed. Retrying after 1000ms
-- Retrying job (attempt 1)
-- Request failed. Retrying after 1000ms
-- Retrying job (attempt 1)
-- Request failed. Retrying after 1000ms
-- Retrying job (attempt 1)
-- Request failed. Retrying after 1000ms
-- Retrying job (attempt 1)
-- Request failed. Retrying after 2000ms
-- Retrying job (attempt 2)
Loading data for 54 collections... done
Updating 80 fields to required... done
Loading 4 dashboards... done
Loading 17 flows... done
Loading settings... done
Loading 0 translations... done
Loading 50 presets... done
Found 1 extensions total: 1 registry extensions (including 0 bundles), and 0 local extensions
All extensions are already installed
Loading extensions... done
------------------
Template applied successfully.
```
---

And when applying to a fresh, local Directus instance ->

```
-- Logged in as Admin User
=== Applying template - test8 to http://localhost:8905

Loading 54 collections and 524 fields... done
Loading 122 relations... done
Loading 5 roles... done
Loading 6 policies... done
Loading 290 permissions... done
Loading 19 users... done
 »   Warning: Directus Error: Invalid foreign key for field "role" in collection "directus_access". (Status: 400) {"collection":"directus_access","field":"role","code":"INVALID_FOREIGN_KEY"}
 »   Context: {"access":{"id":"3107a8ed-18ea-44fa-ac19-866536aac9f0","role":"c8cc6feb-db01-40a5-bbdc-ead36a040115","user":null,"policy":"c8cc6feb-db01-40a5-bbdc-ead36a040115","sort":1},"operation":"createAccess"}
Loading 6 accesses... done
Loading 27 folders... done
Loading 235 files... done
-- Request failed. Retrying after 1000ms
-- Retrying job (attempt 1)
-- Request failed. Retrying after 1000ms
-- Retrying job (attempt 1)
-- Request failed. Retrying after 1000ms
-- Retrying job (attempt 1)
Loading data for 54 collections... done
Updating 80 fields to required... done
Loading 4 dashboards... done
Loading 17 flows... done
Loading settings... done
Loading 0 translations... done
Loading 50 presets... done
Found 1 extensions total: 1 registry extensions (including 0 bundles), and 0 local extensions
-- Server under pressure. Retrying after 5000ms
-- Retrying job (attempt 1)
 »   Warning: Directus Error: Value for field "id" in collection "directus_extensions" has to be unique. (Status: 400) {"collection":"directus_extensions","field":"id","code":"RECORD_NOT_UNIQUE"}
-- Failed to install directus-extension-sync
Installing 1 extensions... done
Finished installing extensions
------------------
Template applied successfully.
```

File: src/lib/load/update-required-fields.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+export default async function updateRequiredFields(dir: string) {
+  const fieldsToUpdate = readFile('fields', dir)
+  .filter(field => field.meta.required === true || field.schema?.is_nullable === false || field.schema?.is_unique === true)
+
+  ux.action.start(ux.colorize(DIRECTUS_PINK, `Updating ${fieldsToUpdate.length} fields to required`))
+
+  for await (const field of fieldsToUpdate) {
+    try {
+      await api.client.request(updateField(field.collection, field.field, {meta: {...field.neta}, schema: {...field.schema}}))
```
Comment (the-other-dev) [inline src/lib/load/update-required-fields.ts:18]: Is `field.neta` a typo and should it be `field.meta` instead?

**Comment by bryantgillespie** [comment]
> Update: I tried `npx directus-template-cli@0.5.0-beta.17 apply` to repeat the agencyOS template installation and it looks like `policies.json` hasn't been updated yet, possibly you do this extraction in CI/CD or something? I'll be looking around for a solve for the next hour. Would very much appreciate pointers for a work around which could make agencyOS usable while you are finalizing this fix :)
> 
> ```
> Loading 76 collections and 700 fields... done
> Loading 176 relations... done
> Loading 4 roles... done
>  ›   Warning: Unknown error: "File not found: 
>  ›   /home/runner/.npm/_npx/081441f1bf72b51a/node_modules/directus-template-cli/dist/lib/downloads/official/agencyos/src/policies.json"
>     Error: ENOENT: no such file or directory, open 
>     '/home/runner/.npm/_npx/081441f1bf72b51a/node_modules/directus-template-cli/dist/lib/downloads/official/agencyos/src/policies.json'
>     Code: ENOENT
>     ```
> ```

The AgencyOS template has not actually been updated yet for V11 so that is the issue.

**Comment by arladmin** [comment]
@alexvdvalk @bryantgillespie 

```
npx directus-template-cli@beta apply
Need to install the following packages:
directus-template-cli@0.5.0-beta.19
...
...
...
-- Logged in as Admin User
=== Applying template - test7 to http://localhost:8905

Loading 54 collections and 524 fields... done
Loading 122 relations... done
Loading 5 roles... done
Loading 6 policies... done
Loading 290 permissions... done
Loading 19 users... done
 »   Warning: Directus Error: Invalid foreign key for field "role" in collection "directus_access". (Status: 400) {"collection":"directus_access","field":"role","code":"INVALID_FOREIGN_KEY"}
 »   Context: {"access":{"id":"3107a8ed-18ea-44fa-ac19-866536aac9f0","role":"c8cc6feb-db01-40a5-bbdc-ead36a040115","user":null,"policy":"c8cc6feb-db01-40a5-bbdc-ead36a040115","sort":1},"operation":"createAccess"}
Loading 6 accesses... done
Loading 27 folders... done
Loading 235 files... done
-- Request failed. Retrying after 1000ms
-- Retrying job (attempt 1)
-- Request failed. Retrying after 1000ms
-- Retrying job (attempt 1)
-- Request failed. Retrying after 1000ms
-- Retrying job (attempt 1)
-- Request failed. Retrying after 1000ms
-- Retrying job (attempt 1)
-- Request failed. Retrying after 1000ms
-- Retrying job (attempt 1)
-- Request failed. Retrying after 1000ms
-- Retrying job (attempt 1)
Loading data for 54 collections... done
Updating 80 fields to required... done
Loading 4 dashboards... done
Loading 17 flows... done
Loading settings... done
Loading 0 translations... done
Loading 50 presets... done
Loading extensions... done
------------------
Template applied successfully.
```

**Comment by arladmin** [comment]
@bryantgillespie 


```
npm run build

> directus-template-cli@0.5.0-beta.19 build
> shx rm -rf dist && tsc -b

node_modules/@directus/types/dist/vue.d.ts:1:26 - error TS2307: Cannot find module 'vue' or its corresponding type declarations.

1 import type { Ref } from 'vue';
                           ~~~~~


Found 1 error.

```

**Comment by alexvdvalk** [comment]
@arladmin Did you run npm i to install all the dependencies? Or maybe it's best to to delete your node_modules and reinstall.

**Comment by arladmin** [comment]
> @arladmin Did you run npm i to install all the dependencies? Or maybe it's best to to delete your node_modules and reinstall.

@alexvdvalk 

```
PS D:\Local Projects\Directus Template CLI\directus-template-cli-11-updates\directus-template-cli-11-updates> npm i git+https://github.com/directus-labs/directus-template-cli.git#v11-updates

added 169 packages, and audited 1944 packages in 2m

194 packages are looking for funding
  run `npm fund` for details

24 vulnerabilities (7 moderate, 17 high)

To address issues that do not require attention, run:
  npm audit fix

Some issues need review, and may require choosing
a different dependency.

Run `npm audit` for details.
PS D:\Local Projects\Directus Template CLI\directus-template-cli-11-updates\directus-template-cli-11-updates> 
PS D:\Local Projects\Directus Template CLI\directus-template-cli-11-updates\directus-template-cli-11-updates> npm run build

> directus-template-cli@0.5.0-beta.19 build
> shx rm -rf dist && tsc -b

node_modules/@directus/types/dist/vue.d.ts:1:26 - error TS2307: Cannot find module 'vue' or its corresponding type declarations.

1 import type { Ref } from 'vue';
                           ~~~~~


Found 1 error.

```

**Comment by arladmin** [comment]
> > @arladmin Did you run npm i to install all the dependencies? Or maybe it's best to to delete your node_modules and reinstall.
> 
> @alexvdvalk
> 
> ```
> PS D:\Local Projects\Directus Template CLI\directus-template-cli-11-updates\directus-template-cli-11-updates> npm i git+https://github.com/directus-labs/directus-template-cli.git#v11-updates
> 
> added 169 packages, and audited 1944 packages in 2m
> 
> 194 packages are looking for funding
>   run `npm fund` for details
> 
> 24 vulnerabilities (7 moderate, 17 high)
> 
> To address issues that do not require attention, run:
>   npm audit fix
> 
> Some issues need review, and may require choosing
> a different dependency.
> 
> Run `npm audit` for details.
> PS D:\Local Projects\Directus Template CLI\directus-template-cli-11-updates\directus-template-cli-11-updates> 
> PS D:\Local Projects\Directus Template CLI\directus-template-cli-11-updates\directus-template-cli-11-updates> npm run build
> 
> > directus-template-cli@0.5.0-beta.19 build
> > shx rm -rf dist && tsc -b
> 
> node_modules/@directus/types/dist/vue.d.ts:1:26 - error TS2307: Cannot find module 'vue' or its corresponding type declarations.
> 
> 1 import type { Ref } from 'vue';
>                            ~~~~~
> 
> 
> Found 1 error.
> ```

---

@alexvdvalk 

I was able to finally build it by updating the `tsconfig.json` ->

image

**Comment by bryantgillespie** [comment]
> At a quick glance, the following didn't get copied/got copied over wrongly:

> Settings (Project Name)
> Appearance (Project Color)

This the expected behavior for v0.5.0 moving forward.  

The primary use case for this tool is to load Directus "starter" templates (created by our team and by our community) for both self-hosted projects and projects hosted on our Cloud.

For project settings, we merge instead of overwrite. We use the existing ones as the defaults, so we don't overwrite the existing Project Name or other settings that may already be configured.

We have many users who apply more than one template to an instance and they were frustrated because the old version reset all their project settings each time based on the incoming template.

> Bookmarks (only 50 out of 80 got copied) (that seems to be a result of "[// Only get the global presets](https://github.com/directus-labs/directus-template-cli/blob/a1df2cf837db87bf8b794b03188210ec500d5c4a/src/lib/extract/extract-presets.ts#L15)" )

Again this is the expected behavior based on our use case.

---

I think your use case is similar but a little different - so we can't guarantee that the tool will have the exact behavior you're looking for. However, this tool is under MIT license. So you're welcome to fork and make whatever changes you need to support your exact use case.

**Comment by sandias42** [comment]
Update: I tried 

> > Update: I tried `npx directus-template-cli@0.5.0-beta.17 apply` to repeat the agencyOS template installation and it looks like `policies.json` hasn't been updated yet, possibly you do this extraction in CI/CD or something? I'll be looking around for a solve for the next hour. Would very much appreciate pointers for a work around which could make agencyOS usable while you are finalizing this fix :)
> > ```
> > Loading 76 collections and 700 fields... done
> > Loading 176 relations... done
> > Loading 4 roles... done
> >  ›   Warning: Unknown error: "File not found: 
> >  ›   /home/runner/.npm/_npx/081441f1bf72b51a/node_modules/directus-template-cli/dist/lib/downloads/official/agencyos/src/policies.json"
> >     Error: ENOENT: no such file or directory, open 
> >     '/home/runner/.npm/_npx/081441f1bf72b51a/node_modules/directus-template-cli/dist/lib/downloads/official/agencyos/src/policies.json'
> >     Code: ENOENT
> >     ```
> > ```
> 
> The AgencyOS template has not actually been updated yet for V11 so that is the issue.

Got it, that makes sense. When do you guess the AgencyOS template will be updated to V11? Or is there some way for me to do that myself? Thanks! (love your videos by the way Bryant :D)

**Comment by beatlecz** [comment]
Any progress here?

**Review by alexvdvalk — APPROVED** [review]

**Pull request merged**